Fix: Don't render empty info unit heading blocks

chosak opened the pull request:
Currently if an info unit contains an empty heading block (meaning empty text), an empty `<div class="m-info-unit_heading-text"><h3></h3></div>` gets rendered on the page:

![image](https://user-images.githubusercontent.com/654645/102130716-29230000-3e1f-11eb-81ae-0b108dbbb35f.png)

If a heading has empty text, it should not be rendered as part of an info unit. This commit removes the branch of logic that does this.

## How to test this PR

To test, visit a page with an empty heading, for example in production:

http://localhost:8000/consumer-tools/educator-tools/adult-financial-education/

Alternatively, create a new e.g. Sublanding page with an info unit group, then add an info unit with an empty heading.

## Notes and todos

Fixes internal platform#3854.

## Checklist

- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)

anselmbradford commented:
This ends up with an image alone inside a link. I think we need to describe it in some way, either by ensuring we have a proper `alt` value set on the image (see https://www.w3.org/WAI/tutorials/images/functional/), or requiring the heading text be entered but have the `u-visually-hidden` class added if it is not intended to be displayed.

Scotchester commented:
Hmm… this is a little more complex than it would appear. See #5630 and #5497. I'm still trying to get my head around what the issue with #5497 was that precipitated #5630.

chosak commented:
> Hmm… this is a little more complex than it would appear.

@Scotchester thanks for catching that complexity. I've addressed this in some additional changes that I've added in a new commit here. The need to check `heading.text` is because of cases where the heading is defined manually instead of from a HeadingBlock, for example when InfoUnits are [rendered manually](https://github.com/cfpb/consumerfinance.gov/blob/e7c3f73cb793c2310cbf8c30a712e43e8208e9af/cfgov/jinja2/v1/_includes/organisms/filterable-list.html#L166-L178) on [Everyone Has a Story](https://www.consumerfinance.gov/consumer-tools/everyone-has-a-story/).

> This ends up with an image alone inside a link.

@anselmbradford in the case where an editor has created an InfoUnit without heading text, if the image has alt text, it will be rendered with that alt text:

![image](https://user-images.githubusercontent.com/654645/102938759-ef837200-447a-11eb-9a56-39e20b32cee4.png)

It's only in the case when the image has no alt text set that we for some reason [use totally different logic](https://github.com/cfpb/consumerfinance.gov/blob/e7c3f73cb793c2310cbf8c30a712e43e8208e9af/cfgov/jinja2/v1/_includes/molecules/info-unit-2.html#L77-L84) to instead render the image using `background-image`.*

However, in the case where no heading text has been set, AND the image has no alt text, we don't really have any good text to use. Would the best behavior would be to set the image alt text to match the link text ("Explore resources" above)?

---

*Strangely, this logic seems to have broken sometime between when I opened this PR a week ago and today. You can see that the [Financial education for adults](https://www.consumerfinance.gov/consumer-tools/educator-tools/adult-financial-education/) page originally mentioned above now no longer seems to render the image, unrelated to this PR:

![image](https://user-images.githubusercontent.com/654645/102938604-9f0c1480-447a-11eb-8bfe-bb2396d321c0.png)

Any idea what might have happened recently here? If we go with the change I suggest above, maybe this becomes a moot point because we remove this code. But I'd like to understand what changed in case it has other impacts elsewhere.

anselmbradford commented:
> Strangely, this logic seems to have broken sometime between when I opened this PR a week ago and today. You can see that the Financial education for adults page originally mentioned above now no longer seems to render the image, unrelated to this PR:

If I go into Wagtail locally and add an alt attribute value for that image, it then shows up. Is it possible this was originally an image and had an `alt` attribute value when you opened the PR, but the `alt` attribute was then removed and it turned into a background image? At any rate, the no-alt-attribute state is broken.

chosak commented:
> the alt attribute was then removed and it turned into a background image

That's right! I forgot that I had noticed the `alt` was incorrect and had asked @csebianlander to remove it, but neither of us noticed that it broke the rendering of the info unit.

anselmbradford commented:
> That's right! I forgot that I had noticed the alt was incorrect and had asked @csebianlander to remove it, but neither of us noticed that it broke the rendering of the info unit.

My guess is the intent here is proper schematics. If the image had no `alt` attribute it is considered "decorative" and can be ignored by screenreaders. However, according to https://www.w3.org/TR/WCAG20-TECHS/H67.html we could set an empty `alt` attribute value for this purpose. It would probably be best to always keep it an `img` to avoid situations like this where it changes drastically based on minor edits in Wagtail.

chosak commented:
@anselmbradford I've added a commit that removes the `<div>` rendering in favor of always using `<img>` with `alt=""` when necessary.

To confirm locally, first visit this URL and note an empty `alt` attribute, because neither the InfoUnit heading nor the image alt text is set:

http://localhost:8000/consumer-tools/educator-tools/adult-financial-education/

Next, edit the page and add heading text to the InfoUnit. You should see the heading show up visually but, deliberately, not get added as alt text, because we don't want to duplicate for screen readers.

Finally, remove that InfoUnit heading text but add alt text to the image. It should get properly added to the image's `alt` attribute.

anselmbradford commented:
> Next, edit the page and add heading text to the InfoUnit. You should see the heading show up visually but, deliberately, not get added as alt text, because we don't want to duplicate for screen readers.
>
> Finally, remove that InfoUnit heading text but add alt text to the image. It should get properly added to the image's alt attribute.

I seems like you should be able to manually add both a heading and alt text, but that doesn't appear possible. When I add a heading the `alt` text is ignored. Is this intentional?

chosak commented:
> I seems like you should be able to manually add both a heading and alt text, but that doesn't appear possible. When I add a heading the alt text is ignored. Is this intentional?

This is [the existing behavior](https://github.com/cfpb/consumerfinance.gov/blob/430f3845e047b5f81b15413dd79554804fbeadb5/cfgov/jinja2/v1/_includes/molecules/info-unit-2.html#L71); note the comment says this is for "Empty alt attribute if header text is present to prevent screen reader repetition".

This was introduced in https://github.com/cfpb/consumerfinance.gov/commit/d1d67e9b359021a396f060bc4877816f95d36a49, and refers to internal https://github.local/CFGOV/platform/issues/2936. @contolini does this logic still make sense to you?

anselmbradford commented:
> This was introduced in d1d67e9, and refers to internal https://github.local/CFGOV/platform/issues/2936. @contolini does this logic still make sense to you?

Hmmm the problem there was that the heading and image could be a link, but unless I'm missing it, that's not possible as infounits currently appear in Wagtail, no? With them not being links I don't see why an `alt` value and heading couldn't both exist.

chosak commented:
It is possible in some situations, for example https://www.consumerfinance.gov/consumer-tools/everyone-has-a-story/ (which would now be different due to the changes in this PR, at http://localhost:8000/consumer-tools/everyone-has-a-story/).

With the new changes this ends up as:

```html
<div class="m-info-unit">
  <a class="m-info-unit_heading-link" href="...">
    <img alt="" class="m-info-unit_image" src="...">
    <div class="m-info-unit_heading-text">
      <h2 class="h3">Ana: Debts you don’t owe</h2>
    </div>
  </a>
  <div class="m-info-unit_content">
    ...
  </div>
</div>
```

If we added the image alt text, you'd have two different pieces of text in the link -- wouldn't this be wrong?

anselmbradford commented:
> If we added the image alt text, you'd have two different pieces of text in the link -- wouldn't this be wrong?

Yes. Can we ignore the `alt` only in the case there's a heading and `alt` text inside a link, but otherwise include both?

chosak commented:
@anselmbradford @Scotchester I've pushed a commit that excludes the image alt text _only_ if the image is being linked _and_ if there is also heading text. So if the image isn't being linked _or_ if it is being linked but there's no heading, the alt text will be included.

This introduces one minor problem, which is that this alt text help text is no longer accurate:

![image](https://user-images.githubusercontent.com/654645/103703532-b288a680-4f75-11eb-86c8-fd3b26566c58.png)

This could be edited to say something like "this text will not appear if the image is linked and if heading text is also provided", but... yuck. Any thoughts on what to do here? (cc @sonnakim)

anselmbradford commented:
> This could be edited to say something like "this text will not appear if the image is linked and if heading text is also provided", but... yuck. Any thoughts on what to do here?

I think the correct solution here would be to handle the behavior the same way in Wagtail as on the frontend. So if the image is a link and there is both alt text and a heading present, then the alt text input should be disabled with the addition of the `disabled` attribute to that input field. How hard would this be? Seems like it may mean getting into Wagtail's JS?

chosak commented:
> How hard would this be? Seems like it may mean getting into Wagtail's JS?

@anselmbradford I totally agree that this would be better. Unfortunately Wagtail doesn't provide an easy way to add this kind of functionality, where fields are shown/hidden in the editor depending on values entered in other fields. In theory we could write some kind of custom JS and use our own widget for the InfoUnit block, but I don't believe we've done this before and am not sure how hard it would be to do. I can ask in the Wagtail Slack #support channel.

I realize that actually the problem is even worse, because the help text is defined on the ImageBasic block, so there's no easy way to customize that text only for images within InfoUnits. So we might be stuck with that text.

Scotchester commented on cfgov/jinja2/v1/_includes/organisms/filterable-list.html:
```diff
@@ -169,7 +165,11 @@
                                         'upload': post.preview_image,
                                         'alt': post.preview_image.alt if post.preview_image.alt else '',
```
Indentation fix, while we're here.

```suggestion
                                            'upload': post.preview_image,
                                            'alt': post.preview_image.alt if post.preview_image.alt else '',
```

Scotchester reviewed (approved):
This is a nice improvement; thank you! I have a few non-blocking suggestions for your consideration.

chosak commented on cfgov/jinja2/v1/_includes/blocks/heading.html:
```diff
@@ -1,6 +1,41 @@
-    {% endif -%}
-    {{ value.text | safe }}
-</{{ value.level }}>
+{# ==========================================================================
+
+   Heading
+
+   ==========================================================================
```
I like this idea but would prefer to do it for a subsequent PR.

chosak commented on cfgov/jinja2/v1/_includes/blocks/heading.html:
```diff
@@ -1,6 +1,41 @@
+   ========================================================================== #}
+{%- macro heading( value ) -%}
+
+{%- set _default_level = 'h3' -%}
```
Rendering the template without `value.level` won't generate a rendering error unless Jinja is configured to use [`StrictUndefined`](https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.StrictUndefined) (which we don't do for consumerfinance.gov). It'll still render the template but put this message in the HTML:

> {{ no such element: dict object['level'] }}

That said, you make a good point, especially since we seem to set a [default level of h2 when a HeadingBlock is used](https://github.com/cfpb/consumerfinance.gov/blob/ef3597113c1fa6317389841eee5318326518db22/cfgov/v1/blocks.py#L147) in Wagtail.

I've added another commit that removes the default level. I've checked everywhere else this template is used and don't believe there are any other uses where we're not specifying a heading. Would you like to give this another review before I merge?

Scotchester replied:
Sounds good. No need to re-review :+1:

chosak commented:
@anselmbradford @Scotchester unfortunately now this introduces an accessibility error on https://www.consumerfinance.gov/consumer-tools/educator-tools/adult-financial-education/ ("Links do not have a discernable name.")

![image](https://user-images.githubusercontent.com/654645/103926747-5776c100-50e7-11eb-9ea9-5595f3d3fccb.png)

In this case:

- The InfoUnitGroup has a heading, but the InfoUnit does not.
- The image has no alt text assigned, either in Wagtail, or in the InfoUnit configuration.

What is the proper fix here? Should I do what I suggested [above](https://github.com/cfpb/consumerfinance.gov/pull/6168#issuecomment-749811565):

> However, in the case where no heading text has been set, AND the image has no alt text, we don't really have any good text to use. Would the best behavior would be to set the image alt text to match the link text ("Explore resources" above)?

That would mean giving the image the alt text of "Explore resources" and would create two different links to the same place with the same text.

Or, instead of alt text, maybe `<a aria-label="Explore resources">` would be more appropriate?

chosak commented:
Per team discussion, we decided: when the InfoUnit has no heading, but does have an image, and the image is linked, we'll set the image `alt` to be equivalent to the link text ("Explore resources"). I'll open another PR for this.
